src/accountsdb/buffer_pool.zig

```diff
@@ -0,0 +1,1175 @@
+        file_id: FileId,
+        /// inclusive
+        file_offset_start: FileOffset,
+        /// exclusive
+        file_offset_end: FileOffset,
+    ) (error{
+        InvalidArgument,
+        OutOfMemory,
+        InvalidKey,
+        CannotResetAlive,
+        CannotOverwriteAliveInfo,
+        OffsetsOutOfBounds,
+    } || std.posix.PReadError)!CachedRead {
+        const frame_indices = try self.computeFrameIndices(
+            file_id,
+            allocator,
+            file_offset_start,
+            file_offset_end,
+        );
+        errdefer allocator.free(frame_indices);
+
+        // update found frames in the LFU (we don't want to evict these in the next loop)
+        for (frame_indices) |f_idx| {
+            if (f_idx == INVALID_FRAME) continue;
+            try self.eviction_lfu.insert(self.frames_metadata, f_idx);
+            if (!self.frames_metadata.rc[f_idx].acquire()) {
+                // frame has no handles, but memory is still valid
+                self.frames_metadata.rc[f_idx].reset();
+            }
+        }
+
+        // fill in invalid frames with file data, replacing invalid frames with
+        // fresh ones.
+        for (0.., frame_indices) |i, *f_idx| {
+            if (f_idx.* != INVALID_FRAME) continue;
+            // INVALID_FRAME => not found, read fresh and populate
+
+            const frame_aligned_file_offset: FileOffset = @intCast((i * FRAME_SIZE) +
+                (file_offset_start - file_offset_start % FRAME_SIZE));
+            std.debug.assert(frame_aligned_file_offset % FRAME_SIZE == 0);
+
+            f_idx.* = blk: while (true) {
+                if (self.free_list.popOrNull()) |free_idx| {
+                    break :blk free_idx;
+                } else {
+                    try self.evictUnusedFrame();
+                }
+            };
+
+            const bytes_read = try file.pread(&self.frames[f_idx.*], frame_aligned_file_offset);
+            try self.overwriteDeadFrameInfo(
+                f_idx.*,
+                file_id,
+                frame_aligned_file_offset,
+                @intCast(bytes_read),
+            );
+        }
+
+        return CachedRead{
+            .buffer_pool = self,
+            .frame_indices = frame_indices,
+            .first_frame_start_offset = @intCast(file_offset_start % FRAME_SIZE),
+            .last_frame_end_offset = @intCast(((file_offset_end - 1) % FRAME_SIZE) + 1),
+        };
+    }
+};
+
+/// TODO: atomics on all index accesses
+pub const FramesMetadata = struct {
+    pub const InQueue = enum(u2) { none, small, main, ghost };
+
+    /// ref count for the frame. For frames that are currently being used elsewhere.
+    rc: []sig.sync.ReferenceCounter,
+
+    /// effectively the inverse of BufferPool.FrameMap, used in order to
+    /// evict keys by their value
+    key: []FileIdFileOffset,
+
+    /// frequency for the S3_FIFO
+    /// Yes, really, only 0, 1, 2, 3.
+    /// Wonder if it would be faster to pack this.
```

Comment:
Mark this as "TODO:" so it's greppable.

Reply:
Removed this comment, it isn't packable faa39cd5c8d496cc1dbac1ac3059d36d39978db8